hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-common/src/main/java/org/apache/hadoop/yarn/server/uam/UnmanagedAMPoolManager.java

```diff
@@ -450,4 +452,52 @@ public void drainUAMHeartbeats() {
       uam.drainHeartbeatThread();
     }
   }
+
+  /**
+   * Complete FinishApplicationMaster interface calls in batches.
+   *
+   * @param request FinishApplicationMasterRequest
+   * @param appId application Id
+   * @return Returns the Map map,
+   *         the key is subClusterId, the value is FinishApplicationMasterResponse
+   */
+  public Map<String, FinishApplicationMasterResponse> batchFinishApplicationMaster(
+      FinishApplicationMasterRequest request, String appId) {
+
+    Map<String, FinishApplicationMasterResponse> responseMap = new HashMap<>();
+    Set<String> subClusterIds = this.unmanagedAppMasterMap.keySet();
+
+    if (subClusterIds != null && !subClusterIds.isEmpty()) {
+      ExecutorCompletionService<Map<String, FinishApplicationMasterResponse>> finishAppService =
+          new ExecutorCompletionService<>(this.threadpool);
+      LOG.info("Sending finish application request to {} sub-cluster RMs", subClusterIds.size());
+
+      for (final String subClusterId : subClusterIds) {
+        finishAppService.submit(() -> {
+          LOG.info("Sending finish application request to RM {}", subClusterId);
+          FinishApplicationMasterResponse uamResponse = null;
+          try {
+            uamResponse = finishApplicationMaster(subClusterId, request);
```

Comment:
```
try {
  FinishApplicationMasterResponse uamResponse = finishApplicationMaster(subClusterId, request);
  return Collections.singletonMap(subClusterId, uamResponse);
} catch (Throwable e) {
  LOG.warn("Failed to finish unmanaged application master: RM address: {} ApplicationId: {}",
      subClusterId, appId, e);
  return Collections.singletonMap(subClusterId, null);
}

Reply:
Thanks for your suggestion, the code looks very good, I will modify it.